actors/builtin/reward/reward_state.go

```diff
@@ -3,305 +3,59 @@ package reward
 import (
 	abi "github.com/filecoin-project/specs-actors/actors/abi"
 	big "github.com/filecoin-project/specs-actors/actors/abi/big"
-	builtin "github.com/filecoin-project/specs-actors/actors/builtin"
-	adt "github.com/filecoin-project/specs-actors/actors/util/adt"
 )
 
-// Fractional representation of NetworkTime with an implicit denominator of (2^MintingInputFixedPoint).
-type NetworkTime = big.Int
-
 // A quantity of space * time (in byte-epochs) representing power committed to the network for some duration.
 type Spacetime = big.Int
 
 type State struct {
-	BaselinePower        abi.StoragePower
-	RealizedPower        abi.StoragePower
-	CumsumBaseline       Spacetime
-	CumsumRealized       Spacetime
-	EffectiveNetworkTime NetworkTime
+	// CumsumBaseline is a target CumsumRealized needs to reach for EffectiveNetworkTime to increase
+	CumsumBaseline Spacetime
+	CumsumRealized Spacetime
 
-	SimpleSupply   abi.TokenAmount // current supply
-	BaselineSupply abi.TokenAmount // current supply
+	// EffectiveNetworkTime is ceiling of EffectiveNetworkTime based on CumsumRealized
+	EffectiveNetworkTime abi.ChainEpoch
 
-	// The reward to be paid in total to block producers, if exactly the expected number of them produce a block.
+	// The reward to be paid in per WinCount to block producers.
 	// The actual reward total paid out depends on the number of winners in any round.
-	// This is computed at the end of the previous epoch, and should really be called ThisEpochReward.
 	ThisEpochReward abi.TokenAmount
-
-	// The count of epochs for which a reward has been paid.
-	// This should equal the number of non-empty tipsets after the genesis, aka "chain height".
-	RewardEpochsPaid abi.ChainEpoch
+	Epoch           abi.ChainEpoch
```

Comment:
Please document, "the epoch for which ThisEpochReward and other calculations have been made".

Is "for which" rather than "in which" correct? I can't quite tell from the code. The calculations are made by cron at the end of one epoch, in order to set the reward for the subsequent one. If "for which", then this value should be zero in genesis state, if "in which" then it should be -1.

Reply:
Thanks for pointing this out, I might have an off by one in here.

Reply:
Genesis is block 0 AFAIK, so "in which" would be 0 and "for which" it would be 1.

Reply:
The terminology is confusing, but right now, as I understand it:
- the genesis block header has epoch `0`, which means the messages it contains execute in epoch `0`
- the "genesis state" is the state _prior to_ epoch `0`
- a block header points to the parent state (usually, the state resulting from messages in the prior epoch)
- ∴ that genesis state was evaluated in epoch `-1` and the messages _in_ the genesis block in epoch `0`
- so at the end of genesis state construction this should have value `0` if this is "for which"; it's the value for the messages _in_ the genesis block, to be earned by the miner _of_ the genesis block
- for genesis state construction, that may require initializing it to `-1` before then calling the computation function to set the first epoch reward (which will bump it to `0`)

A key point of confusion is distinguishing the state prior to the genesis block from the state resulting from executing any messages contained in it.

There's a brief discussion about changing that [here](https://github.com/filecoin-project/specs/issues/874), but never actioned.

Reply:
Yeah, I had an off by one. Now the `Epoch` is *for which epoch*, updated comment too.